File: 37-sudoku-solver/37-sudoku-solver.py
import copy
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        def available(i,j):
            row = set([m for m in board[i] if m != "."])
            col = set([board[k][j] for k in range(9) if board[k][j] != "."])
            block = set()
            for m in range(i//3*3,i//3*3+3):
                for n in range(j//3*3,j//3*3+3):
                    if board[m][n] != ".":
                        block.add(board[m][n])
            final = row.union(col)
            final = final.union(block)
            stack = []
            for i in range(1,10):
                if str(i) not in final:
                    stack.append(i)
            return stack
        undecided = dict()
        n = 0
        for i in range(9):
            for j in range(9):
                if board[i][j] == ".":
                    undecided[n] = (i,j)
                    n = n + 1
        total_num = len(undecided)
        def backtracking(n):
            global res
            if n == total_num:
                return True
            point = undecided[n]
            i, j = point[0],point[1]
            nums = available(i, j)
            if not nums:
                return False
            for num in nums:
                board[i][j] = str(num)
                if backtracking(n+1):
                    return True
                board[i][j] = "."
        backtracking(0)
        return board
                
            
                    
        

File: 37-sudoku-solver/37-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        solve(board);
        return;
    }
    public boolean solve(char[][] board){
        int N = 9;
        
        for(int row = 0; row < N; row++)
            for(int col = 0; col < N; col++){
                if(board[row][col] == '.'){
                    for(int num = 0; num < 9; num++){
                        if(isValidStep(board, row, col, num)){
                            board[row][col] = Integer.toString(num+1).charAt(0);
                            if(solve(board))return true;
                            board[row][col] = '.';
                        }
                        
                    }
                    return false;
                }
            }
        return true;
        
        
    }
    public boolean isValidStep(char[][] board, int row, int col, int num){
        //examine row
        int N = 9;
        char curChar = Integer.toString(num+1).charAt(0);
        for(int i = 0; i < N; i++){
            if (board[row][i] == curChar)return false;
        }
        //examine col
        for(int j = 0; j < N; j++){
            if(board[j][col] == curChar)return false;
        }
        
        
        //examine sub
        for(int m = row/3*3; m < row/3*3+3; m++)
            for(int n = col/3*3; n < col/3*3 + 3; n++){
                if(board[m][n] == curChar)return false;
            }
        return true;
    }
    
    
}
                
            
                    
        